File: pnpm11/exec/commands/src/taskRunState.ts
import crypto from 'node:crypto'
import fs, { type FileHandle } from 'node:fs/promises'
import path from 'node:path'
import util from 'node:util'

import { createHexHash } from '@pnpm/crypto.hash'
import { PnpmError } from '@pnpm/error'
import type { TaskGraph, TaskKey, TaskNode } from '@pnpm/workspace.task-scheduler'
import writeFileAtomic from 'write-file-atomic'

const STATE_VERSION = 1
const STATE_DIR = '.pnpm-task-run-state-v1'
const LATEST_STATE_FILE = 'latest.json'
const PUBLISHED_SUFFIX = '.published'
const FINISHED_SUFFIX = '.finished'
const START_LOCK_DIR = 'start.lock'
const LOCK_OWNER_FILE = 'owner'
const LOCK_POLL_INTERVAL_MS = 50
const LOCK_WAIT_MS = 2_000
const LOCK_ABANDONED_MS = 30_000
const RUN_GENERATION_LENGTH = 12
const RUN_ID = /^[0-9a-f]{12}-[0-9a-f-]{1,115}$/

interface TaskId {
  project: string
  task: string
}

interface TaskIdentity extends TaskId {
  scripts: Array<{ name: string, commands: string[] }>
  requested: boolean
  dependencies: TaskId[]
}

interface InvocationIdentity {
  command: 'run' | 'exec'
  params: string[]
  settings: string[]
  tasks: TaskIdentity[]
}

interface StateHeader {
  version: number
  invocation: string
  run: string
}

interface TaskRecord extends TaskId {
  run: string
}

interface FinishRecord {
  run: string
  finished: true
}

export interface TaskRunStateContextOptions {
  command: InvocationIdentity['command']
  params: string[]
  settings?: string[]
  graph: TaskGraph
  workspaceDir: string
  scriptCommands: (node: TaskNode, script: string) => string[]
}

export interface TaskRunExecutionSettings {
  extraBinPaths?: string[]
  extraEnv?: Record<string, string | undefined>
  modulesDir?: string
  nodeExperimentalPackageMap?: boolean
  nodeOptions?: string
  userAgent?: string
}

export function taskRunExecutionSettings (opts: TaskRunExecutionSettings): string[] {
  const extraEnv = Object.entries(opts.extraEnv ?? {})
    .sort(([left], [right]) => compareStrings(left, right))
    .map(([key, value]) => [key, value ?? null])
  return [
    `extra-bin-paths=${JSON.stringify(opts.extraBinPaths ?? [])}`,
    `extra-env=${JSON.stringify(extraEnv)}`,
    `modules-dir=${opts.modulesDir ?? 'node_modules'}`,
    `node-experimental-package-map=${Boolean(opts.nodeExperimentalPackageMap)}`,
    `node-options=${opts.nodeOptions ?? ''}`,
    `user-agent=${opts.userAgent ?? ''}`,
  ]
}

export class TaskRunStateContext {
  readonly invocation: string
  readonly latestStatePath: string
  private readonly opts: TaskRunStateContextOptions
  private readonly nodeModulesDir: string
  private readonly stateDir: string
  private readonly keysById = new Map<string, TaskKey>()

  constructor (opts: TaskRunStateContextOptions) {
    this.opts = opts
    const identity: InvocationIdentity = {
      command: opts.command,
      params: opts.params,
      settings: [...(opts.settings ?? [])].sort(compareStrings),
      tasks: [...opts.graph].map(([key, node]) => {
        const id = taskId(node, opts.workspaceDir)
        this.keysById.set(taskIdKey(id), key)
        return {
          ...id,
          scripts: node.scripts
            .map((name) => ({ name, commands: opts.scriptCommands(node, name) }))
            .sort(compareScripts),
          requested: node.requested,
          dependencies: node.dependencies
            .map((dependency) => taskId(opts.graph.get(dependency)!, opts.workspaceDir))
            .sort(compareTaskIds),
        }
      }).sort(compareTaskIds),
    }
    this.invocation = createHexHash(JSON.stringify(identity))
    this.nodeModulesDir = path.join(opts.workspaceDir, 'node_modules')
    this.stateDir = path.join(this.nodeModulesDir, STATE_DIR)
    this.latestStatePath = path.join(this.stateDir, LATEST_STATE_FILE)
  }

  async readCompletedTasks (): Promise<Set<TaskKey> | undefined> {
    try {
      if (!await this.validateStateDirectory(false)) return undefined
    } catch (err: unknown) {
      if (isStateUnavailableError(err)) return undefined
      throw err
    }
    let latest: StateHeader
    try {
      latest = JSON.parse(await fs.readFile(this.latestStatePath, 'utf8')) as StateHeader
    } catch (err: unknown) {
      if (util.types.isNativeError(err) && 'code' in err && err.code !== 'ENOENT' && !isStateUnavailableError(err)) throw err
      return undefined
    }
    if (latest.version !== STATE_VERSION || latest.invocation !== this.invocation || !RUN_ID.test(latest.run)) return undefined
    let state: { run: string, finished: boolean }
    try {
      state = await this.newestState(latest.run)
    } catch (err: unknown) {
      if (isStateUnavailableError(err)) return undefined
      throw err
    }
    if (state.finished) return undefined
    const filePath = this.journalPath(state.run)
    let contents: string
    try {
      contents = await fs.readFile(filePath, 'utf8')
    } catch (err: unknown) {
      if (util.types.isNativeError(err) && 'code' in err && (err.code === 'ENOENT' || isStateUnavailableError(err))) return undefined
      throw err
    }
    // A record is committed by its newline; a process killed during append
    // can leave only the final record torn.
    const lines = contents.split('\n')
    lines.pop()
    if (lines.length === 0) return undefined
    let header: StateHeader
    try {
      header = JSON.parse(lines[0]) as StateHeader
    } catch {
      return undefined
    }
    if (header.version !== STATE_VERSION || header.invocation !== this.invocation || header.run !== state.run) return undefined
    const completed = new Set<TaskKey>()
    for (const line of lines.slice(1)) {
      let record: TaskRecord | FinishRecord
      try {
        record = JSON.parse(line) as TaskRecord | FinishRecord
      } catch {
        return undefined
      }
      if (record.run !== header.run) continue
      if (isFinishRecord(record)) return undefined
      const key = this.keysById.get(taskIdKey(record))
      if (key == null) return undefined
      completed.add(key)
    }
    return completed
  }

  async start (completedTasks: ReadonlySet<TaskKey>): Promise<TaskRunState> {
    let run = createRunId(Date.now())
    let filePath = this.journalPath(run)
    let file: FileHandle | undefined
    let journalCreated = false
    let lock: StateStartLock | undefined
    try {
      await this.validateStateDirectory(true)
      lock = await StateStartLock.acquire(path.join(this.stateDir, START_LOCK_DIR))
      if (lock == null) {
        return new TaskRunState(filePath, this.publishedPath(run), this.finishedPath(run), undefined, this.opts.workspaceDir, run, completedTasks)
      }
      run = await this.nextRunId()
      filePath = this.journalPath(run)
      const header: StateHeader = { version: STATE_VERSION, invocation: this.invocation, run }
      const completed = [...completedTasks]
        .map((key): TaskRecord => ({ run, ...taskId(this.opts.graph.get(key)!, this.opts.workspaceDir) }))
        .sort(compareTaskIds)
      const contents = [header, ...completed].map((record) => JSON.stringify(record)).join('\n') + '\n'
      await writeFileAtomic(filePath, contents, { mode: 0o600 })
      journalCreated = true
      file = await fs.open(filePath, 'a')
      if (!await lock.isOwner()) {
        await file.close()
        file = undefined
        await unlinkIfExists(filePath)
        journalCreated = false
        return new TaskRunState(filePath, this.publishedPath(run), this.finishedPath(run), undefined, this.opts.workspaceDir, run, completedTasks)
      }
      await writeFileAtomic(this.latestStatePath, JSON.stringify(header), { mode: 0o600 })
      await writeFileAtomic(this.publishedPath(run), '', { mode: 0o600 })
      await this.cleanupOlderFinishedState(run).catch(() => {})
    } catch (err: unknown) {
      await file?.close().catch(() => {})
      if (journalCreated) await unlinkIfExists(filePath).catch(() => {})
      await unlinkIfExists(this.publishedPath(run)).catch(() => {})
      if (isStateUnavailableError(err)) {
        return new TaskRunState(filePath, this.publishedPath(run), this.finishedPath(run), undefined, this.opts.workspaceDir, run, completedTasks)
      }
      throw err
    } finally {
      await lock?.release()
    }
    return new TaskRunState(filePath, this.publishedPath(run), this.finishedPath(run), file, this.opts.workspaceDir, run, completedTasks)
  }

  private journalPath (run: string): string {
    return path.join(this.stateDir, `${this.invocation}.${run}.jsonl`)
  }

  private publishedPath (run: string): string {
    return path.join(this.stateDir, `${this.invocation}.${run}${PUBLISHED_SUFFIX}`)
  }

  private finishedPath (run: string): string {
    return path.join(this.stateDir, `${this.invocation}.${run}${FINISHED_SUFFIX}`)
  }

  private async newestState (latestRun: string): Promise<{ run: string, finished: boolean }> {
    let newestRun = latestRun
    const prefix = `${this.invocation}.`
    const names = new Set(await fs.readdir(this.stateDir))
    let finished = names.has(`${prefix}${latestRun}${FINISHED_SUFFIX}`)
    for (const name of names) {
      if (!name.startsWith(prefix)) continue
      let run: string
      let candidateFinished: boolean
      if (name.endsWith(FINISHED_SUFFIX)) {
        run = name.slice(prefix.length, -FINISHED_SUFFIX.length)
        candidateFinished = true
      } else if (name.endsWith('.jsonl')) {
        run = name.slice(prefix.length, -'.jsonl'.length)
        if (!names.has(`${prefix}${run}${PUBLISHED_SUFFIX}`)) continue
        candidateFinished = false
      } else {
        continue
      }
      if (!RUN_ID.test(run)) continue
      if (runGeneration(run) > runGeneration(newestRun)) {
        newestRun = run
        finished = candidateFinished
      } else if (run === newestRun && candidateFinished) {
        finished = true
      }
    }
    return { run: newestRun, finished }
  }

  private async nextRunId (): Promise<string> {
    let newestGeneration = Date.now().toString(16).padStart(RUN_GENERATION_LENGTH, '0')
    try {
      const latest = JSON.parse(await fs.readFile(this.latestStatePath, 'utf8')) as StateHeader
      if (latest.invocation === this.invocation && RUN_ID.test(latest.run)) {
        newestGeneration = maxString(newestGeneration, runGeneration(latest.run))
      }
    } catch (err: unknown) {
      if (util.types.isNativeError(err) && 'code' in err && err.code !== 'ENOENT') throw err
    }
    const prefix = `${this.invocation}.`
    for (const name of await fs.readdir(this.stateDir)) {
      if (!name.startsWith(prefix)) continue
      const suffix = name.endsWith('.jsonl') ? '.jsonl' : name.endsWith(FINISHED_SUFFIX) ? FINISHED_SUFFIX : undefined
      if (suffix == null) continue
      const run = name.slice(prefix.length, -suffix.length)
      if (RUN_ID.test(run)) newestGeneration = maxString(newestGeneration, runGeneration(run))
    }
    return createRunId(Number.parseInt(newestGeneration, 16) + 1)
  }

  private async cleanupOlderFinishedState (run: string): Promise<void> {
    const prefix = `${this.invocation}.`
    const generation = runGeneration(run)
    const removals: Array<Promise<void>> = []
    for (const name of await fs.readdir(this.stateDir)) {
      if (!name.startsWith(prefix)) continue
      if (!name.endsWith(FINISHED_SUFFIX)) continue
      const olderRun = name.slice(prefix.length, -FINISHED_SUFFIX.length)
      if (!RUN_ID.test(olderRun) || runGeneration(olderRun) >= generation) continue
      removals.push(unlinkIfExists(path.join(this.stateDir, name)).catch(() => {}))
    }
    await Promise.all(removals)
  }

  private async validateStateDirectory (create: boolean): Promise<boolean> {
    if (!await validateRealDirectory(this.nodeModulesDir, create)) return false
    return validateRealDirectory(this.stateDir, create)
  }
}

function createRunId (generation: number): string {
  return `${generation.toString(16).padStart(RUN_GENERATION_LENGTH, '0')}-${crypto.randomUUID()}`
}

function runGeneration (run: string): string {
  return run.slice(0, RUN_GENERATION_LENGTH)
}

function maxString (left: string, right: string): string {
  return left > right ? left : right
}

function isFinishRecord (record: TaskRecord | FinishRecord): record is FinishRecord {
  return 'finished' in record && record.finished
}

export class TaskRunState {
  readonly filePath: string
  private readonly publishedPath: string
  private readonly finishedPath: string
  private readonly file: FileHandle | undefined
  private readonly workspaceDir: string
  private readonly run: string
  private readonly completedTasks: Set<TaskKey>
  private pendingWrite: Promise<void> = Promise.resolve()
  private closePromise: Promise<void> | undefined
  private disabled: boolean

  constructor (
    filePath: string,
    publishedPath: string,
    finishedPath: string,
    file: FileHandle | undefined,
    workspaceDir: string,
    run: string,
    completedTasks: ReadonlySet<TaskKey>
  ) {
    this.filePath = filePath
    this.publishedPath = publishedPath
    this.finishedPath = finishedPath
    this.file = file
    this.workspaceDir = workspaceDir
    this.run = run
    this.completedTasks = new Set(completedTasks)
    this.disabled = file == null
  }

  async recordPassed (key: TaskKey, node: TaskNode): Promise<void> {
    const file = this.file
    if (this.disabled || file == null || this.completedTasks.has(key)) return
    this.completedTasks.add(key)
    const record: TaskRecord = { run: this.run, ...taskId(node, this.workspaceDir) }
    const line = `${JSON.stringify(record)}\n`
    let unavailable = false
    const write = this.pendingWrite.then(async () => {
      if (this.disabled) return
      try {
        await file.appendFile(line)
      } catch (err: unknown) {
        if (!isStateUnavailableError(err)) throw err
        this.disabled = true
        unavailable = true
      }
    })
    this.pendingWrite = write.catch(() => {})
    try {
      await write
    } catch (err: unknown) {
      this.completedTasks.delete(key)
      throw err
    }
    if (unavailable) {
      await this.close().catch(() => {})
      await unlinkIfExists(this.filePath).catch(() => {})
      await unlinkIfExists(this.publishedPath).catch(() => {})
    }
  }

  async finish (): Promise<void> {
    if (this.file == null || this.disabled) return
    if (this.closePromise == null) {
      const finishRecord: FinishRecord = { run: this.run, finished: true }
      const write = this.pendingWrite.then(async () => this.file!.appendFile(`${JSON.stringify(finishRecord)}\n`))
      this.pendingWrite = write.catch(() => {})
      try {
        await write
      } catch (err: unknown) {
        if (!isStateUnavailableError(err)) throw err
      }
    }
    await this.close()
    try {
      await writeFileAtomic(this.finishedPath, '', { mode: 0o600 })
    } catch (err: unknown) {
      if (isStateUnavailableError(err)) return
      throw err
    }
    try {
      await unlinkIfExists(this.publishedPath)
      await unlinkIfExists(this.filePath)
    } catch (err: unknown) {
      if (!isStateUnavailableError(err)) throw err
    }
  }

  async close (): Promise<void> {
    const file = this.file
    if (file == null) return
    this.closePromise ??= this.pendingWrite.then(async () => file.close())
    await this.closePromise
  }
}

class StateStartLock {
  private readonly lockPath: string
  private readonly token: string

  private constructor (
    lockPath: string,
    token: string
  ) {
    this.lockPath = lockPath
    this.token = token
  }

  static async acquire (lockPath: string): Promise<StateStartLock | undefined> {
    return StateStartLock.acquireUntil(lockPath, Date.now() + LOCK_WAIT_MS)
  }

  private static async acquireUntil (lockPath: string, deadline: number): Promise<StateStartLock | undefined> {
    try {
      await fs.mkdir(lockPath)
      const token = `${process.pid}-${Date.now()}-${crypto.randomUUID()}`
      try {
        await fs.writeFile(path.join(lockPath, LOCK_OWNER_FILE), token, { mode: 0o600 })
      } catch (err: unknown) {
        await fs.rm(lockPath, { force: true, recursive: true }).catch(() => {})
        throw err
      }
      return new StateStartLock(lockPath, token)
    } catch (err: unknown) {
      if (!(util.types.isNativeError(err) && 'code' in err && err.code === 'EEXIST')) throw err
    }
    const stats = await fs.lstat(lockPath).catch(() => undefined)
    if (stats == null || stats.isSymbolicLink() || !stats.isDirectory()) return undefined
    if (Date.now() - stats.mtimeMs > LOCK_ABANDONED_MS) {
      const removed = await fs.rm(lockPath, { force: true, recursive: true }).then(() => true, () => false)
      if (removed) return StateStartLock.acquireUntil(lockPath, deadline)
    }
    if (Date.now() >= deadline) return undefined
    await new Promise(resolve => setTimeout(resolve, LOCK_POLL_INTERVAL_MS))
    return StateStartLock.acquireUntil(lockPath, deadline)
  }

  async isOwner (): Promise<boolean> {
    try {
      return await fs.readFile(path.join(this.lockPath, LOCK_OWNER_FILE), 'utf8') === this.token
    } catch (err: unknown) {
      if (util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT') return false
      throw err
    }
  }

  async release (): Promise<void> {
    if (!await this.isOwner().catch(() => false)) return
    await fs.rm(this.lockPath, { force: true, recursive: true }).catch(() => {})
  }
}

async function validateRealDirectory (dir: string, create: boolean): Promise<boolean> {
  let stats
  try {
    stats = await fs.lstat(dir)
  } catch (err: unknown) {
    if (!(util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT')) throw err
    if (!create) return false
    try {
      await fs.mkdir(dir)
    } catch (mkdirErr: unknown) {
      if (!(util.types.isNativeError(mkdirErr) && 'code' in mkdirErr && mkdirErr.code === 'EEXIST')) throw mkdirErr
    }
    stats = await fs.lstat(dir)
  }
  if (stats.isSymbolicLink() || !stats.isDirectory()) {
    throw new PnpmError('UNSAFE_TASK_RUN_STATE_PATH', `Refusing to use task run state directory at "${dir}" because it is a symbolic link or not a directory`)
  }
  return true
}

function taskId (node: TaskNode, workspaceDir: string): TaskId {
  const relative = path.relative(workspaceDir, node.project)
  return {
    project: relative === '' ? '.' : relative.replaceAll(path.sep, '/'),
    task: node.taskName,
  }
}

function taskIdKey (id: TaskId): string {
  return `${id.project}\0${id.task}`
}

function compareTaskIds (left: TaskId, right: TaskId): number {
  return compareStrings(left.project, right.project) || compareStrings(left.task, right.task)
}

function compareScripts (left: { name: string }, right: { name: string }): number {
  return compareStrings(left.name, right.name)
}

function compareStrings (left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0
}

async function unlinkIfExists (filePath: string): Promise<void> {
  try {
    await fs.unlink(filePath)
  } catch (err: unknown) {
    if (!(util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT')) throw err
  }
}

function isStateUnavailableError (err: unknown): boolean {
  return util.types.isNativeError(err) && 'code' in err &&
    (err.code === 'EACCES' || err.code === 'EPERM' || err.code === 'EROFS')
}

File: pnpm/crates/cli/src/cli_args/task_run_state.rs
use journal::{current_generation, remove_state_file, validate_real_directory};
use miette::{IntoDiagnostic, WrapErr as _};
use pnpm_crypto_hash::create_hex_hash;
use pnpm_workspace_task_scheduler::{TaskGraph, TaskKey, TaskNode};
use serde::{Deserialize, Serialize};
use std::{
    collections::{HashMap, HashSet},
    ffi::{OsStr, OsString},
    fs,
    fs::{File, OpenOptions},
    io,
    io::Write as _,
    path::{Path, PathBuf},
    sync::{
        Mutex,
        atomic::{AtomicU64, Ordering},
    },
    time::{Duration, SystemTime, UNIX_EPOCH},
};

const STATE_VERSION: u8 = 1;
const STATE_DIR: &str = ".pnpm-task-run-state-v1";
const LATEST_STATE_FILE: &str = "latest.json";
const PUBLISHED_SUFFIX: &str = ".published";
const FINISHED_SUFFIX: &str = ".finished";
const START_LOCK_DIR: &str = "start.lock";
const LOCK_WAIT: Duration = Duration::from_secs(2);
const LOCK_ABANDONED_AFTER: Duration = Duration::from_secs(30);
const RUN_GENERATION_LENGTH: usize = 12;
static RUN_SEQUENCE: AtomicU64 = AtomicU64::new(0);

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
struct TaskId {
    project: String,
    task: String,
}

#[derive(Serialize)]
struct ScriptIdentity {
    name: String,
    commands: Vec<String>,
}

#[derive(Serialize)]
struct TaskIdentity {
    project: String,
    task: String,
    scripts: Vec<ScriptIdentity>,
    requested: bool,
    dependencies: Vec<TaskId>,
}

#[derive(Serialize)]
struct InvocationIdentity<'a> {
    command: &'a str,
    params: &'a [String],
    settings: &'a [String],
    tasks: Vec<TaskIdentity>,
}

#[derive(Serialize, Deserialize)]
struct StateHeader {
    version: u8,
    invocation: String,
    run: String,
}

#[derive(Serialize, Deserialize)]
struct TaskRecord {
    run: String,
    project: String,
    task: String,
}

#[derive(Serialize, Deserialize)]
struct FinishRecord {
    run: String,
    finished: bool,
}

#[derive(Deserialize)]
#[serde(untagged)]
enum JournalRecord {
    Task(TaskRecord),
    Finish(FinishRecord),
}

pub struct TaskRunExecutionSettings<'a> {
    pub extra_bin_paths: &'a [PathBuf],
    pub extra_env: &'a HashMap<String, String>,
    pub modules_dir: &'a Path,
    pub node_experimental_package_map: bool,
    pub node_options: Option<&'a str>,
    pub user_agent: &'a str,
}

pub fn task_run_execution_settings(opts: &TaskRunExecutionSettings<'_>) -> Vec<String> {
    let extra_bin_paths: Vec<String> =
        opts.extra_bin_paths.iter().map(|path| path.to_string_lossy().into_owned()).collect();
    let mut extra_env: Vec<(&String, &String)> = opts.extra_env.iter().collect();
    extra_env.sort_by_key(|(key, _)| *key);
    vec![
        format!(
            "extra-bin-paths={}",
            serde_json::to_string(&extra_bin_paths).expect("extra bin paths serialize")
        ),
        format!(
            "extra-env={}",
            serde_json::to_string(&extra_env).expect("extra environment serializes")
        ),
        format!("modules-dir={}", opts.modules_dir.to_string_lossy()),
        format!("node-experimental-package-map={}", opts.node_experimental_package_map),
        format!("node-options={}", opts.node_options.unwrap_or_default()),
        format!("user-agent={}", opts.user_agent),
    ]
}

pub struct TaskRunStateContext {
    state_dir: PathBuf,
    latest_state_path: PathBuf,
    invocation: String,
    keys_by_id: HashMap<TaskId, TaskKey>,
    ids_by_key: HashMap<TaskKey, TaskId>,
}

impl TaskRunStateContext {
    pub fn new(
        command: &str,
        params: &[String],
        settings: &[String],
        graph: &TaskGraph,
        workspace_dir: &Path,
        script_commands: impl Fn(&TaskNode, &str) -> Vec<String>,
    ) -> Self {
        let mut keys_by_id = HashMap::with_capacity(graph.len());
        let mut ids_by_key = HashMap::with_capacity(graph.len());
        let tasks: Vec<TaskIdentity> = graph
            .iter()
            .map(|(key, node)| {
                let id = task_id(node, workspace_dir);
                keys_by_id.insert(id.clone(), key.clone());
                ids_by_key.insert(key.clone(), id.clone());
                task_identity(node, id, graph, workspace_dir, &script_commands)
            })
            .collect();
        let invocation = invocation_hash(command, params, settings, tasks);
        let state_dir = workspace_dir.join("node_modules").join(STATE_DIR);
        let latest_state_path = state_dir.join(LATEST_STATE_FILE);
        Self { state_dir, latest_state_path, invocation, keys_by_id, ids_by_key }
    }

    pub fn read_completed_tasks(&self) -> miette::Result<Option<HashSet<TaskKey>>> {
        let Some(latest_run) = self.resumable_latest_run()? else { return Ok(None) };
        let (run, finished) = match self.newest_state(&latest_run) {
            Ok(state) => state,
            Err(error) if error.is_unavailable() => return Ok(None),
            Err(error) => return Err(error.into_report()),
        };
        if finished {
            return Ok(None);
        }
        let file_path = self.journal_path(&run);
        let contents = match fs::read(&file_path) {
            Ok(contents) => contents,
            Err(error)
                if error.kind() == io::ErrorKind::NotFound
                    || is_state_unavailable_error(&error) =>
            {
                return Ok(None);
            }
            Err(error) => {
                return Err(error)
                    .into_diagnostic()
                    .wrap_err_with(|| format!("reading {}", file_path.display()));
            }
        };
        Ok(self.completed_from_journal(&contents, &run))
    }

    pub fn start(&self, completed_tasks: &HashSet<TaskKey>) -> miette::Result<TaskRunState> {
        let mut completed: Vec<&TaskId> =
            completed_tasks.iter().map(|key| &self.ids_by_key[key]).collect();
        completed.sort();
        let (file_path, run, file) = match self.start_file(&completed) {
            Ok(state) => state,
            Err(error) if error.is_unavailable() => {
                let run = run_id(current_generation());
                (self.journal_path(&run), run, None)
            }
            Err(error) => return Err(error.into_report()),
        };
        Ok(TaskRunState {
            file_path,
            published_path: self.published_path(&run),
            finished_path: self.finished_path(&run),
            writer: Mutex::new(TaskRunStateWriter {
                file,
                run,
                completed: completed_tasks.clone(),
            }),
        })
    }

    fn journal_path(&self, run: &str) -> PathBuf {
        self.state_dir.join(format!("{}.{run}.jsonl", self.invocation))
    }

    fn published_path(&self, run: &str) -> PathBuf {
        self.state_dir.join(format!("{}.{run}{PUBLISHED_SUFFIX}", self.invocation))
    }

    fn finished_path(&self, run: &str) -> PathBuf {
        self.state_dir.join(format!("{}.{run}{FINISHED_SUFFIX}", self.invocation))
    }

    fn validate_state_directory(&self, create: bool) -> Result<bool, StateStorageError> {
        let node_modules_dir = self.state_dir.parent().expect("task state directory has a parent");
        if !validate_real_directory(node_modules_dir, create)? {
            return Ok(false);
        }
        validate_real_directory(&self.state_dir, create)
    }
}

pub struct TaskRunState {
    file_path: PathBuf,
    published_path: PathBuf,
    finished_path: PathBuf,
    writer: Mutex<TaskRunStateWriter>,
}

struct TaskRunStateWriter {
    file: Option<File>,
    run: String,
    completed: HashSet<TaskKey>,
}

impl TaskRunState {
    pub fn record_passed(
        &self,
        key: &TaskKey,
        node: &TaskNode,
        workspace_dir: &Path,
    ) -> miette::Result<()> {
        let mut writer = self.writer.lock().expect("task state lock is not poisoned");
        if writer.file.is_none() {
            return Ok(());
        }
        if !writer.completed.insert(key.clone()) {
            return Ok(());
        }
        let id = task_id(node, workspace_dir);
        let record = TaskRecord { run: writer.run.clone(), project: id.project, task: id.task };
        let line = serde_json::to_string(&record).expect("task record serializes");
        let result = writeln!(
            writer.file.as_mut().expect("unfinished task state has an open file"),
            "{line}",
        );
        if let Err(error) = result {
            if is_state_unavailable_error(&error) {
                writer.file.take();
                drop(writer);
                let _ = fs::remove_file(&self.file_path);
                let _ = fs::remove_file(&self.published_path);
                return Ok(());
            }
            writer.completed.remove(key);
            return Err(error)
                .into_diagnostic()
                .wrap_err_with(|| format!("writing {}", self.file_path.display()));
        }
        Ok(())
    }

    pub fn finish(&self) -> miette::Result<()> {
        let mut writer = self.writer.lock().expect("task state lock is not poisoned");
        let Some(mut file) = writer.file.take() else {
            return Ok(());
        };
        let finish_record = FinishRecord { run: writer.run.clone(), finished: true };
        let line = serde_json::to_string(&finish_record).expect("finish record serializes");
        if let Err(error) = writeln!(file, "{line}")
            && !is_state_unavailable_error(&error)
        {
            return Err(error)
                .into_diagnostic()
                .wrap_err_with(|| format!("writing {}", self.file_path.display()));
        }
        drop(file);
        drop(writer);
        if let Err(error) = pnpm_fs::write_atomic(&self.finished_path, &[]) {
            if is_state_unavailable_error(&error) {
                return Ok(());
            }
            return Err(error)
                .into_diagnostic()
                .wrap_err_with(|| format!("writing {}", self.finished_path.display()));
        }
        if !remove_state_file(&self.published_path)? {
            return Ok(());
        }
        remove_state_file(&self.file_path)?;
        Ok(())
    }
}

fn task_id(node: &TaskNode, workspace_dir: &Path) -> TaskId {
    let relative = pnpm_fs::relative_path(workspace_dir, &node.project);
    let project = if relative.as_os_str().is_empty() {
        ".".to_string()
    } else {
        relative.to_string_lossy().replace(std::path::MAIN_SEPARATOR, "/")
    };
    TaskId { project, task: node.task_name.clone() }
}

fn task_identity(
    node: &TaskNode,
    id: TaskId,
    graph: &TaskGraph,
    workspace_dir: &Path,
    script_commands: &impl Fn(&TaskNode, &str) -> Vec<String>,
) -> TaskIdentity {
    let mut scripts: Vec<ScriptIdentity> = node
        .scripts
        .iter()
        .map(|name| ScriptIdentity { name: name.clone(), commands: script_commands(node, name) })
        .collect();
    scripts.sort_by(|left, right| left.name.cmp(&right.name));
    let mut dependencies: Vec<TaskId> = node
        .dependencies
        .iter()
        .map(|dependency| task_id(&graph[dependency], workspace_dir))
        .collect();
    dependencies.sort();
    TaskIdentity {
        project: id.project,
        task: id.task,
        scripts,
        requested: node.requested,
        dependencies,
    }
}

/// The invocation's identity hash over its command line and the sorted
/// settings and tasks.
fn invocation_hash(
    command: &str,
    params: &[String],
    settings: &[String],
    mut tasks: Vec<TaskIdentity>,
) -> String {
    tasks.sort_by(|left, right| {
        left.project.cmp(&right.project).then_with(|| left.task.cmp(&right.task))
    });
    let mut settings = settings.to_vec();
    settings.sort();
    let identity =
        serde_json::to_string(&InvocationIdentity { command, params, settings: &settings, tasks })
            .expect("task invocation identity serializes");
    create_hex_hash(&identity)
}

enum StateStorageError {
    Io { error: io::Error, operation: &'static str, path: PathBuf },
    UnsafePath(PathBuf),
}

impl StateStorageError {
    fn io(error: io::Error, operation: &'static str, path: &Path) -> Self {
        Self::Io { error, operation, path: path.to_path_buf() }
    }

    fn is_unavailable(&self) -> bool {
        matches!(self, Self::Io { error, .. } if is_state_unavailable_error(error))
    }

    fn into_report(self) -> miette::Report {
        match self {
            Self::Io { error, operation, path } => {
                let result: miette::Result<()> = Err(error)
                    .into_diagnostic()
                    .wrap_err_with(|| format!("{operation} {}", path.display()));
                result.expect_err("task state storage error cannot succeed")
            }
            Self::UnsafePath(path) => {
                let path = path.display();
                miette::miette!(
                    code = "ERR_PNPM_UNSAFE_TASK_RUN_STATE_PATH",
                    "Refusing to use task run state directory at {} because it is a symbolic link or not a directory",
                    path,
                )
            }
        }
    }
}

fn is_state_unavailable_error(error: &io::Error) -> bool {
    matches!(error.kind(), io::ErrorKind::PermissionDenied | io::ErrorKind::ReadOnlyFilesystem)
}

fn run_id(generation: u64) -> String {
    let sequence = RUN_SEQUENCE.fetch_add(1, Ordering::Relaxed);
    format!("{generation:012x}-{}-{sequence}", std::process::id())
}

#[cfg(test)]
mod tests;

fn initial_journal_contents(header: &StateHeader, completed: &[&TaskId]) -> String {
    let mut contents = serde_json::to_string(header).expect("task state header serializes");
    contents.push('\n');
    for id in completed {
        let record = TaskRecord {
            run: header.run.clone(),
            project: id.project.clone(),
            task: id.task.clone(),
        };
        contents.push_str(&serde_json::to_string(&record).expect("task record serializes"));
        contents.push('\n');
    }
    contents
}

/// Remove an unpublished journal if it cannot be reopened for appending.
fn open_journal_for_append(file_path: &Path) -> Result<File, StateStorageError> {
    OpenOptions::new().append(true).open(file_path).map_err(|error| {
        let _ = fs::remove_file(file_path);
        StateStorageError::io(error, "opening", file_path)
    })
}

mod journal;
